tests/atomic_fence/atomic_fence.cpp

```diff
@@ -0,0 +1,250 @@
+                                         sycl::memory_scope scope,
+                                         const std::string& scope_name) {
+// FIXME: re-enable when sycl::info::device::atomic_fence_scope_capabilities is
+// implemented
+#if !SYCL_CTS_COMPILING_WITH_DPCPP
+  auto scopes =
+      queue.get_device()
+          .get_info<sycl::info::device::atomic_fence_scope_capabilities>();
+  if (std::find(scopes.begin(), scopes.end(), scope) == orders.end()) {
+    WARN(std::string("Device does not support atomic_fence with scope = ") +
+         scope_name);
+    return false;
+  }
+#endif
+  return true;
+}
+
+/**
+ * @brief Factory function for getting type_pack with memory_order values for
+ * atomic_fence
+ */
+inline auto get_memory_orders() {
+  static const auto memory_orders =
+      value_pack<sycl::memory_order, sycl::memory_order::release,
+                 sycl::memory_order::acq_rel,
+                 sycl::memory_order::seq_cst>::generate_named();
+  return memory_orders;
+}
+
+/**
+ * @brief Factory function for getting type_pack with test_type values for
+ * atomic_fence
+ */
+inline auto get_test_types() {
+  static const auto test_types =
+      value_pack<test_type, test_type::single_group,
+                 test_type::between_groups>::generate_named();
+  return test_types;
+}
+
+/**
+ * @brief Factory function for getting type_pack with memory_scope for
+ * atomic_fence inside single work-group
+ */
+inline auto get_memory_scopes_single_group() {
+  static const auto memory_scopes =
+      value_pack<sycl::memory_scope, sycl::memory_scope::work_group,
+                 sycl::memory_scope::device,
+                 sycl::memory_scope::system>::generate_named();
+  return memory_scopes;
+}
+
+/**
+ * @brief Factory function for getting type_pack with memory_scope for
+ * atomic_fence between work-grous
+ */
+inline auto get_memory_scopes_between_work_groups() {
+  static const auto memory_scopes =
+      value_pack<sycl::memory_scope, sycl::memory_scope::device,
+                 sycl::memory_scope::system>::generate_named();
+  return memory_scopes;
+}
+
+bool check_memory_order_scope_capabilities(sycl::queue& queue,
+                                           sycl::memory_order order,
+                                           sycl::memory_scope scope,
+                                           const std::string& order_name,
+                                           const std::string& scope_name) {
+  if (!check_atomic_fence_order_capability(queue, order, order_name)) {
+    return false;
+  }
+  if (sycl::memory_order::release == order) {
+    if (!check_atomic_fence_order_capability(queue, sycl::memory_order::acquire,
+                                             "memory_order::acquire")) {
+      return false;
+    }
+  }
+
+  if (!check_atomic_fence_scope_capability(queue, scope, scope_name)) {
+    return false;
+  }
+  return true;
+}
+
+template <typename OrderT, typename ScopeT, typename TestT>
```

Comment:
Have you noticed that this test we spent a lot of time to devise is very complex but has 0 comment? ;-)
You need to comment the tricky parts, why it is designed this way, etc.
Otherwise, in a few years from know, we will not remember why it is designed this way.

Reply:
@keryell Comments to source code added.
Should I add the same explanation to the source code, as I added to the test plan?

Reply:
Yes.

Reply:
Explanation added.